### PR TITLE
Fix support for generic types in Tree component

### PR DESCRIPTION
**Pull request by jjhong922**
#### Changes proposed in this pull request:

- Fix support for generic types in Tree component

addresses https://github.com/palantir/blueprint/issues/2627

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/blueprint](https://github.com/palantir/blueprint), @jjhong922! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: packages/core/src/components/tree/tree.tsx [hunk 1]
```diff
@@ -80,15 +92,16 @@ export class Tree<T = {}> extends React.Component<ITreeProps<T>, {}> {
         return this.nodeRefs[nodeId];
     }
 
-    private renderNodes(treeNodes: ITreeNode[], currentPath?: number[], className?: string): JSX.Element {
+    private renderNodes(treeNodes: Array<ITreeNode<T>>, currentPath?: number[], className?: string): JSX.Element {
```
Comment (giladgray) [inline packages/core/src/components/tree/tree.tsx]: is this change necessary? i don't think it is as this component doesn't care about the `<T>` type: it's just for the consumer. let's try to keep this as minimal as possible.

File: packages/core/src/components/tree/tree.tsx [hunk 2]
```diff
@@ -80,15 +92,16 @@ export class Tree<T = {}> extends React.Component<ITreeProps<T>, {}> {
...
         if (treeNodes == null) {
             return null;
         }
 
         const nodeItems = treeNodes.map((node, i) => {
             const elementPath = currentPath.concat(i);
+            const TypedTreeNode = TreeNode.ofType<T>();
```
Comment (giladgray) [inline packages/core/src/components/tree/tree.tsx]: shouldn't _have_ to do this. can you revert this safely?

File: packages/core/src/components/tree/tree.tsx [hunk 3]
```diff
@@ -63,6 +67,14 @@ export class Tree<T = {}> extends React.Component<ITreeProps<T>, {}> {
 
     private nodeRefs: { [nodeId: string]: HTMLElement } = {};
 
+    public nodeFromPath(path: number[]): ITreeNode<T> {
+        let treeNodes = this.props.contents;
+        for (let _i = 0; _i < path.length - 1; _i++) {
```
Comment (giladgray) [inline packages/core/src/components/tree/tree.tsx]: call it `i` cuz it's used. leading `_` is convention for _unused but necessary_ variables.

File: packages/core/src/components/tree/tree.tsx [hunk 4]
```diff
@@ -63,6 +67,14 @@ export class Tree<T = {}> extends React.Component<ITreeProps<T>, {}> {
 
     private nodeRefs: { [nodeId: string]: HTMLElement } = {};
 
+    public nodeFromPath(path: number[]): ITreeNode<T> {
```
Comment (giladgray) [inline packages/core/src/components/tree/tree.tsx]: i'm confused, where was this function defined before?

Comment (giladgray) [reply packages/core/src/components/tree/tree.tsx]: also please put this after `render()`. render should always be the first public method in a react component.

File: packages/core/src/components/tree/tree.tsx (hunk 1 above)
Comment (jjhong922) [reply packages/core/src/components/tree/tree.tsx]: Since the method is called from the render method which passes in `this.props.contents`, the renderNodes function has to take in typed interfaces to render typed corresponding TypedTreeNodes. The problem that existed was that a TypedTree rendered nodes that would not be able to take in nodeData of the passed in type as expected.

File: packages/core/src/components/tree/tree.tsx (hunk 2 above)
Comment (jjhong922) [reply packages/core/src/components/tree/tree.tsx]: This change follows the renderNodes argument type change. I believe some sort of Typed Tree Node is necessary, but I'm not sure if there is a cleaner way.

File: packages/core/src/components/tree/tree.tsx (hunk 4 above)
Comment (jjhong922) [reply packages/core/src/components/tree/tree.tsx]: This function was not defined before, I just made it as a typed alternative to the static nodeFromPath. I can remove it if you don't think it's necessary though.

Comment (giladgray) [reply packages/core/src/components/tree/tree.tsx]: oh yes please remove. the instance method is far less preferable than the static one because it requires grabbing a ref.

File: packages/core/src/components/tree/tree.tsx (hunk 1 above)
Comment (giladgray) [reply packages/core/src/components/tree/tree.tsx]: 🤔 the tree nodes are rendered internally by the tree so we don't actually need the type information inside the component (it's only useful in event handlers).

Comment (jjhong922) [reply packages/core/src/components/tree/tree.tsx]: But since renderNodes is the function that passes down the onClick, onContextMenu, etc. handlers, it cannot create default TreeNodes unless the event handlers match up.

File: packages/core/src/components/tree/tree.tsx [hunk 5]
```diff
@@ -71,6 +75,14 @@ export class Tree<T = {}> extends React.Component<ITreeProps<T>, {}> {
         );
     }
 
+    public nodeFromPath(path: number[]): ITreeNode<T> {
```
Comment (giladgray) [inline packages/core/src/components/tree/tree.tsx]: but actually remove this. we already have the static method.

Comment (jjhong922) [reply packages/core/src/components/tree/tree.tsx]: oh sounds good, didn't see the hidden comment :P

**Comment by jjhong922** [comment]
@giladgray Is the test that failed a flaky one? Is it possible you could rerun CI?

**Comment by giladgray** [comment]
thanks @jjhong922!
